crates/services/gas_price_service/src/fuel_gas_price_updater/fuel_da_source_adapter.rs

```diff
@@ -1,16 +1,125 @@
 use crate::fuel_gas_price_updater::{
-    DaCommitDetails,
-    DaCommitSource,
+    DaBlockCosts,
+    Error::CouldNotFetchDARecord,
+    GetDaBlockCosts,
     Result as GasPriceUpdaterResult,
 };
+use anyhow::anyhow;
+use std::sync::Arc;
+use tokio::sync::Mutex;
 
-#[derive(Default, Clone)]
-pub struct FuelDaSource;
+pub mod block_committer_costs;
+pub mod dummy_costs;
+pub mod service;
 
-impl DaCommitSource for FuelDaSource {
-    fn get_da_commit_details(
-        &mut self,
-    ) -> GasPriceUpdaterResult<Option<DaCommitDetails>> {
-        todo!() // TODO(#2139): pending research on how to get the data from the block committer
+pub use block_committer_costs::BlockCommitterDaBlockCosts;
+pub use dummy_costs::DummyDaBlockCosts;
+pub use service::*;
+
+pub const POLLING_INTERVAL_MS: u64 = 10_000;
+
+pub type DaBlockCostsProvider = Arc<Mutex<Option<DaBlockCosts>>>;
+
+impl GetDaBlockCosts for DaBlockCostsProvider {
+    fn get(&mut self) -> GasPriceUpdaterResult<Option<DaBlockCosts>> {
+        let mut da_block_costs_guard = self.try_lock().map_err(|err| {
+            CouldNotFetchDARecord(anyhow!(
+                "Failed to lock shared gas price state: {:?}",
+                err
+            ))
+        })?;
+
+        let da_block_costs = da_block_costs_guard.clone();
+
+        // now mark it as consumed because we don't want to serve the same data
+        // multiple times
+        *da_block_costs_guard = None;
```

Comment:
I would suggest implementing the DA source service as a stream that guarantees the uniqueness of events. It is the responsibility of the gas provider to decide what to do with these unique events.

Reply:
okay, agreed ~ I'll use a channel for this purpose. I don't think there is much value in adding the stream to the sources (block committer, dummy etc) because they are relatively simple in logic, and don't need to know that we are polling the endpoints at regular intervals.

Reply:
uniqueness handled in 8ff00388e